### PR TITLE
Finds first code cell for imports

### PR DESCRIPTION
**Pull request by tnwei**
Pull request to address #27

### CODE REVIEW

File: src/pyforest/static/utils.js [hunk 1]
```diff
@@ -1,14 +1,23 @@
 define([], function () {
+	window.findFirstCodeCell = function findFirstCodeCell() {
+		var cells = Jupyter.notebook.get_cells();
+		  for (var i = 0; i < cells.length; i++) {
```
Comment (8080labs) [inline src/pyforest/static/utils.js:4]: Should have the same indentation as the line before

File: src/pyforest/static/utils.js [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
+			  if (cells[i]["cell_type"] == "code" ){
+				  return i;
+			  }
+		  }
+	};
+	
 	function setup_lab(notebookTracker) {
 		window._pyforest_update_imports_cell = function (imports_string) {
-			var cell = notebookTracker.currentWidget.model.cells.get(0).value;
+			var cell = notebookTracker.currentWidget.model.cells.get(findFirstCodeCell()).value;
```
Comment (8080labs) [inline src/pyforest/static/utils.js:13]: JupyterLab needs a different logic - the above logic only works in Jupyter Notebook

File: src/pyforest/static/utils.js [hunk 3]
```diff
@@ -1,14 +1,23 @@
 define([], function () {
+	window.findFirstCodeCell = function findFirstCodeCell() {
```
Comment (8080labs) [inline src/pyforest/static/utils.js:2]: Try to use the new function without adding it to window because we don't want to spam the users window functions

**Review by 8080labs — CHANGES_REQUESTED** [review]
Thank you for the PR! :)
Please have a look at the comment and change accordingly. If something is unclear, please let us know

File: src/pyforest/static/utils.js (hunk 1 above)
Comment (8080labs) [reply src/pyforest/static/utils.js:4]: I mean the diff between `var cells` and the `for` block

Comment (tnwei) [reply src/pyforest/static/utils.js:4]: Aha right thanks for pointing this out

File: src/pyforest/static/utils.js (hunk 3 above)
Comment (tnwei) [reply src/pyforest/static/utils.js:2]: Removed from window namespace

File: src/pyforest/static/utils.js (hunk 2 above)
Comment (tnwei) [reply src/pyforest/static/utils.js:13]: Been looking at this for a while, can you give me some tips on what I should be focused on? A bit lost navigating the differences between Notebook and Lab.

Comment (8080labs) [reply src/pyforest/static/utils.js:13]: Sorry, but I also did not work with this so far. You might try to get all the cells and then read the type for each cell. Similar to the Jupyter Notebook logic. The code is just a little bit different for the notebook. Biggest challenge might be to debug this in order to inspect whats available.

You might look at this here and then go from there:
https://jupyterlab.readthedocs.io/en/stable/developer/notebook.html#structure-of-the-notebook-plugin

https://jupyterlab.github.io/jupyterlab/cells/classes/cellmodel.html

Comment (tnwei) [reply src/pyforest/static/utils.js:13]: I believe the building blocks required sit here: https://jupyterlab.github.io/jupyterlab/cells/globals.html#iscodecellmodel, just need to find a way to get this into the code. Kept running into JLab build issues, so this is going to take a while.

Comment (8080labs) [reply src/pyforest/static/utils.js:13]: Alright, thank you for letting us know and hang in there - we know that struggle! 👍

Comment (tkrabel) [reply src/pyforest/static/utils.js:13]: FYI: I'm taking a hit on this myself. Already have something working for JupyterLab. PR coming soon.

Comment (tkrabel) [reply src/pyforest/static/utils.js:13]: Solved with PR #46

**Comment by tkrabel** [comment]
This can be closed now. I added JupyterLab support myself in PR #46.

**Pull request closed without merging**